paragonie/password_lock
# Version 4.0

**NOTE THIS PR CONTAINS HUGE BC BREAKS**

- Add PasswordHasherInterface and default PasswordHasher implementation 
    - Add ability to modify the hashing algorithm 
    - Add ability to change hashing options 
    - Add ability to use different hashing methods ( not recommended ) 

- More Test
    - Test PasswordHasher implementation 
    - Test Null-Byte Truncation
    - Test Key rotation 
- Set minimum PHP version to 7.1 
- Remove legacy support
- fixes #10 
- refactor methods name for cleaner interface `lock`/`check` is cleaner and easier to type than `hashAndEncrypt`/`decryptAndVerify` 
- pass the key as a constructor argument ( you don't have to keep moving the key around the application ) 

example : 

```php
<?php

use Defuse\Crypto\Key;
use ParagonIE\PasswordLock\{
    PasswordLock,
    Hasher\PasswordHasher
};

// use Argon2I algorithm instead of Bcrypt
$hasher = new PasswordHasher(PASSWORD_ARGON2I, [
    'memory_cost' => 2048
]);

$key = Key::createNewRandomKey();

$password = new PasswordLock($key, $hasher);

$storeMe = $password->lock('password');

assert( $password->check('password', $storeMe) );

```
